### PR TITLE
Merge to single chart directory, add deletion job for uninstall

### LINKED ISSUE #59: Combine the k8ssandra and k8ssandra-cluster charts
The separate, two chart installation is awkward and error-prone.

Initially the k8ssandra chart was supposed to install the stack - Prometheus, Grafana, Reaper, cass-operator, prometheus-operator, etc. The k8ssandra-cluster chart was supposed to be focused on creating and configuring the CassandraDatacenter. We intended for everything or as much as possible in the k8ssandra chart to be cluster-scoped by default. This means you would only run a single instances of cass-operator, Prometheus, Grafana, etc.

We ran into various issues with trying to make things cluster-scoped; consequently, more and more things have been moved from the k8ssandra chart to the k8ssandra-cluster chart. To be clear, we intend to support cluster-scoped deployments of the various resources. We simply did not have time to address those issues prior to KubeCon.

As it currently stands, k8ssandra-cluster only installs cass-operator and prometheus-operator.

There was another reason for the separate charts. cass-operator adds a [finalizer](https://book.kubebuilder.io/reference/using-finalizers.html) to every CassandraDatacenter. Kubernetes will not delete an object until all finalizers are cleared. If cass-operator is deleted before the CassandraDatacenter, deletion is essentially blocked. You have to manually removed the finalizer in order for deletion of the CassandraDatacenter to complete. I mention all of this because there is no ordered deletion with `helm uninstall`. If everything is in the same chart, we could potentially wind up in a band state where deletion of the CassandraDatacenter is blocked.

We can work around the unordered deletion with a `pre-delete` hook in the chart. (See this [section](https://helm.sh/docs/topics/charts_hooks/#hooks-and-the-release-lifecycle) in the Helm docs for details). I will write up a separate ticket for implementing the `pre-delete` hook.

Provided we have a way to cleanly handle deletions, I propose the following for this issue:

* Move all components/templates into the k8ssandra chart
* Do a namespace-scoped install of cass-operator by default
* Do a namespace-scoped install of prometheus-operator by default

With these changes everything will be namespace-scoped by default. In future tickets we can look to add support multi-namespace or cluster-wide scope for the various components.

### PR DESCRIPTION
**Pull request by burmanm**
Merge k8ssandra-cluster and k8ssandra charts, add pre-delete deletion job to provide uninstall.

Fixes #59 and #60

### CODE REVIEW

**Comment by burmanm** [comment]
This should be added to the tests in the integration test PRs eventually.

File: pkg/cleaner/cleaner.go
Comment (jsanda) [inline pkg/cleaner/cleaner.go:43]: This is going to delete all CassandraDatacenters in the namespace which could include ones that were not install by the Helm chart. We need to be careful and only delete ones installed by the chart. We probably need to filter release label.

File: pkg/go.mod
Comment (jsanda) [inline pkg/go.mod:1]: I think I prefer to have a single `go.mod` for the whole repo.

File: charts/k8ssandra/templates/cleaner/role.yaml
Comment (jsanda) [inline charts/k8ssandra/templates/cleaner/role.yaml:2]: I would like to refine the role to only include what is actually needed.

File: charts/k8ssandra/templates/cleaner/role.yaml
Comment (jsanda) [inline charts/k8ssandra/templates/cleaner/role.yaml:36]: I am pretty sure we do not need anything from `monitoring.coreos.com`

File: charts/k8ssandra/templates/cleaner/role.yaml
Comment (jsanda) [inline charts/k8ssandra/templates/cleaner/role.yaml:56]: Is this needed?

File: charts/k8ssandra/templates/cleaner/role.yaml
Comment (jsanda) [inline charts/k8ssandra/templates/cleaner/role.yaml:67]: Can the verbs here be limited to get, list, and delete? Might need watch as well.

File: charts/k8ssandra/templates/cleaner/role.yaml
Comment (jsanda) [inline charts/k8ssandra/templates/cleaner/role.yaml:23]: Is this needed?

File: charts/k8ssandra/templates/cleaner/role.yaml
Comment (jsanda) [inline charts/k8ssandra/templates/cleaner/role.yaml:29]: Are any of these needed?

**Review by jsanda — CHANGES_REQUESTED** [review]

File: pkg/go.mod
Comment (burmanm) [reply pkg/go.mod:1]: This will be easier to add after the tests has been merged.

**Comment by burmanm** [comment]
@jsanda I cherry-picked those two commits.

File: pkg/go.mod
Comment (jsanda) [reply pkg/go.mod:1]: 👍

File: pkg/cleaner/cleaner.go
Comment (jsanda) [inline pkg/cleaner/cleaner.go:16]: Should these be constants?

File: charts/k8ssandra/templates/cleaner/role.yaml
Comment (jsanda) [inline charts/k8ssandra/templates/cleaner/role.yaml:15]: Is `update` needed?

**Review by jsanda — CHANGES_REQUESTED** [review]
I still need to spend some time testing the changes,  but this looks good! 

We are going to have to do some doc updates as well. I can help with that if you want.

File: pkg/cleaner/cleaner_test.go
Comment (jeffbanks) [inline pkg/cleaner/cleaner_test.go:1]: @burmanm @jsanda 
Providing a [link](https://github.com/k8ssandra/k8ssandra/tree/operator-integration-tests) over to the integration testing foundations underway.   

As part of those iterations, there are quite a few lessons-learned and functions available that we could reference for much of this cleanup.

I was hopeful we could get that PR merged before we got too far in the other areas.  As such, I would be willing to assist with a ticket to get that effort merged with this cleanup test as we move forward.

File: pkg/main.go
Comment (jeffbanks) [inline pkg/main.go:15]: Is this something that the developers will run to perform a local cleanup?

File: pkg/cleaner/suite_test.go
Comment (jeffbanks) [inline pkg/cleaner/suite_test.go:33]: Same comment as above about integration test foundations PR reuse.

**Review by jeffbanks — APPROVED** [review]
Overall, LGTM.  I left 3 comments w/ some suggestions and one question about the main.  Other than those, I agree w/ John's feedback for the additional changes.

File: pkg/main.go
Comment (burmanm) [reply pkg/main.go:15]: This is ran as part of the ``helm uninstall`` 's pre-delete hook

File: pkg/cleaner/cleaner_test.go
Comment (burmanm) [reply pkg/cleaner/cleaner_test.go:1]: I'm not following really, what would this benefit from? It only creates two CassandraDatacenters to the testenv and then runs the cleaner basically.

Comment (jeffbanks) [reply pkg/cleaner/cleaner_test.go:1]: The thought is we would have more sameness/consistency in our k8ssandra tests with build-up and reuse of integration test functions (e.g. util module). Maybe something similar for our unit testing.

[Here](https://github.com/k8ssandra/k8ssandra/blob/operator-integration-tests/tests/integration/util/lib.go) is the current version of the `util module` where I've been building some common operations.

I included some setup & helper models as well to manage the various options (operator, cluster, network, etc.) that  many of the integration tests require.

Comment (burmanm) [reply pkg/cleaner/cleaner_test.go:1]: Sure, but none of those are really useful for this test. This test isn't running against real cluster.

Comment (jsanda) [reply pkg/cleaner/cleaner_test.go:1]: @jeffbanks The test use [envtest](https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg/envtest). As @burmanm, it is not a real cluster. envtest does deploy etcd and the apiserver, but that is it.

**Comment by burmanm** [comment]
.. don't merge, the rebase broke something.

**Comment by burmanm** [comment]
Closing this in favor of newer PR (safer way to do it).

**Pull request closed without merging**
